src/role.rs

```diff
@@ -35,28 +45,74 @@ impl<'a> Role<'a> {
     {
         let wrapped: Software = Box::new(move || Box::pin(software()));
 
-        Self::Simulated {
+        Self {
+            role: RoleType::Simulated { software: wrapped },
             rt,
-            software: wrapped,
-            handle,
+            handle: Some(handle),
         }
     }
 
-    pub(crate) fn now(&self) -> Instant {
-        let rt = match self {
-            Role::Client { rt, .. } => rt,
-            Role::Simulated { rt, .. } => rt,
-        };
+    pub(crate) fn is_client(&self) -> bool {
+        matches!(self.role, RoleType::Client)
+    }
 
-        rt.now()
+    /// Runtime's current time
+    pub(crate) fn now(&self) -> Instant {
+        self.rt.now()
     }
 
     pub(crate) fn tick(&self, duration: Duration) {
-        let rt = match self {
-            Role::Client { rt, .. } => rt,
-            Role::Simulated { rt, .. } => rt,
-        };
+        self.rt.tick(duration);
+    }
+
+    /// Returns true if task was finished
+    pub(crate) fn is_finished(&self) -> bool {
+        self.handle.as_ref().map_or(true, JoinHandle::is_finished)
+    }
 
-        rt.tick(duration)
+    /// Checks finished runtime for errors.
+    ///
+    /// Consumes the handle to extract task's execution result.
+    /// No-op if errors were already pulled or still task is still running
+    pub(crate) fn finished_err(&mut self) -> Result<()> {
```

Comment:
Discussed a bit offline... perhaps something like:

```rust
fn is_finished(&mut self) -> Result<bool>
```

Reply:
This idea didn't work out well because we need to filter host with running software for loop in the `step` function. I've moved logic to extract error and drop handle into `Role::tick`.